loam/evaluation.lisp

```diff
@@ -47,166 +47,231 @@
+	     ;; signal
+	     (rule (,unhasher (tag-address ,tag) digest) <--
+	       (ingress ptr) (when (has-tag-p ptr ,tag)) (ptr-value ptr digest))
+
+	     ;; signal
+	     (rule ,@alloc-forms <--
+	       (,unhasher (tag-address ,tag) digest)
+	       (,hash-rel ,@unhash-args digest))
+
+	     ;; signal
+	     (rule (,hasher (tag-address ,tag) ,@hash-args) <--
+	       (egress ,name)
+	       (,name-rel ,@signal-args ,name)
+	       ,@digest<-mem-forms)
+
+	     ;; signal
+	     (rule ,@egress-forms <--
+	       (egress ,name) (,name-rel ,@signal-args ,name))
+	     )))))
+
+  ;; Digest only memory. FIXME: Need to define this.
+  ;; Example: sym, builtin.
+  (defun defmem-digest-only (prog-name superclasses config)
+    (let* ((name (getf config :name))
+	   (tag (getf config :tag))
+	   (initial-addr (getf config :initial-addr))
+	   (digest-mem (concat-sym name '-digest-mem))
+	  )
+     `(progn
+       (defprogram ,prog-name ,superclasses
+	 (include ptr-program) ;; TODO: Maybe this should not be hardcoded.
+
+	 (relation (,digest-mem wide dual-element))
+
+	 (rule (,digest-mem value (alloc ,tag (dual ,initial-addr))) <--
+	   (alloc (tag-address ,tag) value))
+	 
+	 (rule (ptr-value ,name value) <--
+	   (,digest-mem value addr) (let ((,name (ptr ,tag (dual-value addr))))))))))
+
+  ;; Immediate memory, for which the pointer value can be represented by exactly the wide value.
+  ;; Thus, instead of allocating memory and managing ptr-value mappings, we can just translate
+  ;; directly when ingressing and egressing.
+  ;;
+  ;; Examples: num, err
+  (defun defmem-immediate (prog-name superclasses config)
+    (let ((name (getf config :name))
+	  (tag (getf config :tag)))
+     `(progn
+       (defprogram ,prog-name ,superclasses
+	 (include ptr-program) ;; TODO: Maybe this should not be hardcoded.
+
+	 ;; When ingressing (a.k.a alloc), just copy the wide value into the ptr.
+	 (rule (ptr-value ptr value) <--
+	   (alloc tag value)
+	   (when (is-tag-p tag ,tag))
+	   (let ((ptr (ptr ,tag (wide-nth 0 value))))))
+
+	 ;; When egressing, just copy the ptr vale into the wide.
+	 (rule (ptr-value ptr value) <--
+	   (egress ptr)
+	   (when (has-tag-p ptr ,tag))
+	   (let ((value (widen (ptr-value ptr))))))))))
+  )
 
-  ;; signal
-  (rule (hash4 body-tag body-value closed-env-tag closed-env-value) <--
-    (egress thunk)
-    (thunk-rel body closed-env thunk)
-    (let ((body-tag (widen (ptr-tag body)))
-	  (closed-env-tag (widen (ptr-tag closed-env)))))
-    (ptr-value body body-value) (ptr-value closed-env closed-env-value)))
-
-(defprogram fun-mem (hash-cache)
+;; FIXME: Annoyingly, I have to copy this over to get the macro expansion to work.
+;; When importing, we end up using the symbols shadowed under the allocation package,
+;; for example `allocation::ingress`, and this interferes with execution.
+;; There should definitely be a way to prevent this from happening in the first place,
+;; but I haven't been able to get anything to work.
+;;
```

Comment:
I'm not totally following the issue, but it's not worth getting into the weeds on now, so let's just live with this comment.